<a href="https://colab.research.google.com/github/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/blob/main/nec_Mar_28_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install geopandas
!pip install wheel
!pip install numpy
!pip install shapely
!pip install fiona
!pip install pyproj
!pip install pandas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import io
import requests

In [5]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv')

Viewing original number of rows and columns

In [6]:
# Count the number of rows and columns
num_rows = df.shape[0]
num_columns = df.shape[1]

# Print the results
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 23021
Number of columns: 31


Removing rows 2-8064, before metrics such as Jitter, Latency, and Packet loss were reliably measured, and counting the new number of rows

In [7]:
# Read the CSV file
df = pd.read_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv')

# Remove rows 2 through 8064
df.drop(df.index[1:8064], inplace=True)

# Save the updated DataFrame
df.to_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv', index=False)

# Count the number of rows
num_rows = df.shape[0]

# Print the results
print("Number of rows:", num_rows)

Number of rows: 14958


Looking at the % of missing items in columns

In [8]:
# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                       0.000000
TEST DATE / UTC HOUR          0.000000
GEO LATITUDE                  0.000000
GEO LONGITUDE                 0.000000
DOWNLOAD SPEED                0.000000
UPLOAD SPEED                  0.000000
JITTER                        0.006685
LATENCY                       0.006685
PACKET LOSS                   0.006685
MOBILE TEST                   0.000000
ISP                           0.000000
CLIENT IP                     0.000000
CLIENT CITY                   0.000000
CLIENT PROVINCE               0.000000
CLIENT POSTAL CODE            0.000000
AS NUMBER                     0.006685
IPV6 SUPPORTED                0.000000
DNSSEC SUPPORTED              0.000000
TEST SERVER                   0.000000
browser_test_url              0.006685
CONTRACTED DOWNLOAD SPEED    72.850648
CONTRACTED UPLOAD SPEED      76.868565
CONTRACTED DATA CAP          84.616927
CONTRACTED MONTHLY COST      77.343228
SERVICE TYPE                 71.653964
USER SATISFACTION        

Removing columns with greater than 66% missing values, and then listing the remaining number of rows & columns

In [9]:
# List of columns to remove
columns_to_remove = [
    'CONTRACTED DOWNLOAD SPEED',
    'CONTRACTED UPLOAD SPEED',
    'CONTRACTED DATA CAP',
    'CONTRACTED MONTHLY COST',
    'SERVICE TYPE',
    'USER SATISFACTION',
    'CONNECTION TYPE',
    'USER COMMENTS',
    'HEATMAP_PROJECT_AREA'
]

# Remove the columns
df.drop(columns=columns_to_remove, inplace=True)

# Save the updated DataFrame
df.to_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv', index=False)

# Count the number of rows and columns
num_rows = df.shape[0]
num_columns = df.shape[1]

# Print the results
print("Number of rows:", num_rows)
print("Number of columns:", num_columns)

Number of rows: 14958
Number of columns: 22


Checking for duplicates in the TEST ID column

In [10]:
# Count the number of duplicated TEST ID values
num_duplicates = df.duplicated(subset='TEST ID', keep=False).sum()

# Print the number of duplicates
print(f"There are {num_duplicates} rows with duplicated TEST ID values.")

There are 0 rows with duplicated TEST ID values.


Looking for duplicate rows

In [11]:
# Find duplicate rows based on all columns
duplicates = df[df.duplicated()]

# Display the duplicate rows (if any)
if len(duplicates) > 0:
    print(f"Found {len(duplicates)} duplicate rows:")
    print(duplicates)
else:
    print("No duplicate rows found.")

No duplicate rows found.


Now that we have removed the columns of missing data, re-checking the % of remaining missing data

In [12]:
# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                  0.000000
TEST DATE / UTC HOUR     0.000000
GEO LATITUDE             0.000000
GEO LONGITUDE            0.000000
DOWNLOAD SPEED           0.000000
UPLOAD SPEED             0.000000
JITTER                   0.006685
LATENCY                  0.006685
PACKET LOSS              0.006685
MOBILE TEST              0.000000
ISP                      0.000000
CLIENT IP                0.000000
CLIENT CITY              0.000000
CLIENT PROVINCE          0.000000
CLIENT POSTAL CODE       0.000000
AS NUMBER                0.006685
IPV6 SUPPORTED           0.000000
DNSSEC SUPPORTED         0.000000
TEST SERVER              0.000000
browser_test_url         0.006685
HEATMAP_MUNICIPALITY     0.000000
HEATMAP_CENSUS_TRACTS    0.000000
dtype: float64


Removing rows then confirming the % of missing items per column.

In [13]:
# List of columns to check for missing values
columns_to_check = ['JITTER', 'LATENCY', 'PACKET LOSS', 'AS NUMBER', 'browser_test_url']

# Remove rows with missing values in specified columns
df.dropna(subset=columns_to_check, inplace=True)

# Save the updated DataFrame
df.to_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv', index=False)

# Calculate the percentage of missing cells in each column
missing_percentages = df.isna().sum() / len(df) * 100

# Print the result
print(missing_percentages)

TEST ID                  0.0
TEST DATE / UTC HOUR     0.0
GEO LATITUDE             0.0
GEO LONGITUDE            0.0
DOWNLOAD SPEED           0.0
UPLOAD SPEED             0.0
JITTER                   0.0
LATENCY                  0.0
PACKET LOSS              0.0
MOBILE TEST              0.0
ISP                      0.0
CLIENT IP                0.0
CLIENT CITY              0.0
CLIENT PROVINCE          0.0
CLIENT POSTAL CODE       0.0
AS NUMBER                0.0
IPV6 SUPPORTED           0.0
DNSSEC SUPPORTED         0.0
TEST SERVER              0.0
browser_test_url         0.0
HEATMAP_MUNICIPALITY     0.0
HEATMAP_CENSUS_TRACTS    0.0
dtype: float64


We had 14958 rows, let's see how many were removed

In [14]:
# Count the number of rows
num_rows = df.shape[0]

# Print the results
print("Number of rows:", num_rows)

Number of rows: 14956


Only 2 rows removed. Skipping standardizing ISP

Looking at for upload/download speeds of 0 mbps

In [15]:
# Count the number of cells equal to zero in the 'DOWNLOAD SPEED' column
download_speed_zeros = (df['DOWNLOAD SPEED'] == 0).sum()

# Count the number of cells equal to zero in the 'UPLOAD SPEED' column
upload_speed_zeros = (df['UPLOAD SPEED'] == 0).sum()

# Print the counts
print("Number of cells equal to zero in 'DOWNLOAD SPEED' column:", download_speed_zeros)
print("Number of cells equal to zero in 'UPLOAD SPEED' column:", upload_speed_zeros)

Number of cells equal to zero in 'DOWNLOAD SPEED' column: 0
Number of cells equal to zero in 'UPLOAD SPEED' column: 7


I now need to remove the rows that have upload speeds of 0, and then list how many cells = 0 to confirm removal was successful

In [16]:
# Count the number of cells equal to zero in the 'UPLOAD SPEED' column before removal
upload_speed_zeros_before = (df['UPLOAD SPEED'] == 0).sum()

# Remove rows where UPLOAD SPEED is equal to 0
df = df[df['UPLOAD SPEED'] != 0]

# Count the number of cells equal to zero in the 'UPLOAD SPEED' column after removal
upload_speed_zeros_after = (df['UPLOAD SPEED'] == 0).sum()

# Save the updated DataFrame
df.to_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv', index=False)

# Print the count before and after removal
print("Number of cells equal to zero in 'UPLOAD SPEED' column before removal:", upload_speed_zeros_before)
print("Number of cells equal to zero in 'UPLOAD SPEED' column after removal:", upload_speed_zeros_after)

Number of cells equal to zero in 'UPLOAD SPEED' column before removal: 7
Number of cells equal to zero in 'UPLOAD SPEED' column after removal: 0


Standarding the all the sub-categories of Winnipeg to be just "Winnipeg"

In [17]:
df['CLIENT CITY'] = df['CLIENT CITY'].str.replace('.*Winnipeg.*', 'Winnipeg', regex=True)

# Get the unique client cities
unique_cities = df['CLIENT CITY'].unique()

# Save the updated DataFrame
df.to_csv('https://raw.githubusercontent.com/InternetSocietyManitobaChapter/JupyterNotebook_NorthEndConnect/main/nec%20Mar%2028%202024.csv', index=False)

# Print the unique client cities
for city in unique_cities:
    print(city)

Winnipeg
Toronto
Vancouver
La Salle
Oakbank
Marathon
Cambridge
Edmonton
Georgetown
Port Hope
Thorold
Old Toronto
Blackstock
Selkirk
Lindsay
Charlemagne
Steinbach
London
Etobicoke
Brandon
Marchand
Baie-d'Urfe
Saint-Quentin
Stonewall
Barrie
Markdale
Hartland
Kenora
Montreal
Calgary
Dowling
Bradford
Dryden
Regina
Thunder Bay
Sydney
North York
Brampton
Mississauga
Heritage Pointe
Cobourg
Hamilton
Cooks Creek
Moose Jaw
Teulon
Headingley
Hague
Markham
Chestermere
Peterborough
Ottawa
Cornwall
Saskatoon
Virden
Sault Ste. Marie


We want to remove any rows where LATENCY >= 30,000 and then list how many rows were removed

In [18]:
# Count the number of rows before removing
count_before = len(df)

# Remove rows where latency >= 30,000
df = df[df['LATENCY'] < 30000]

# Count the number of rows after removing
count_after = len(df)

# Calculate the count of removed rows
count_removed = count_before - count_after

# Print the count of removed rows
print("Number of removed rows:", count_removed)

Number of removed rows: 1


Descriptive stats for DOWNLOAD SPEED for Winnipeg

In [19]:
# Filter the DataFrame to include only rows where CLIENT CITY is Winnipeg
filtered_df = df[df['CLIENT CITY'] == 'Winnipeg']

# Extract the DOWNLOAD SPEED column as a Series
download_speed = df['DOWNLOAD SPEED']

# Calculate the mean and median
mean = download_speed.mean()
median = download_speed.median()

# Calculate other statistical measures
range_val = download_speed.max() - download_speed.min()
std_dev = download_speed.std()
skewness = download_speed.skew()

# Find the minimum and maximum values
min_download_speed = download_speed.min()
max_download_speed = download_speed.max()

# Print the results
print("Min DOWNLOAD SPEED:", min_download_speed)
print("Max DOWNLOAD SPEED:", max_download_speed)
print("Mean DOWNLOAD SPEED:", mean)
print("Median DOWNLOAD SPEED:", median)
print("Range DOWNLOAD SPEED:", range_val)
print("Standard Deviation DOWNLOAD SPEED:", std_dev)
print("Skewness DOWNLOAD SPEED:", skewness)

Min DOWNLOAD SPEED: 0.01
Max DOWNLOAD SPEED: 1071.58
Mean DOWNLOAD SPEED: 113.46242641156009
Median DOWNLOAD SPEED: 65.50999999999999
Range DOWNLOAD SPEED: 1071.57
Standard Deviation DOWNLOAD SPEED: 129.87363613804794
Skewness DOWNLOAD SPEED: 1.8582255603211209


Descriptive stats for JITTER for Winnipeg

In [20]:
# Filter the DataFrame to include only rows where CLIENT CITY is Winnipeg
filtered_df = df[df['CLIENT CITY'] == 'Winnipeg']

# Extract the JITTER column as a Series
jitter = df['JITTER']

# Calculate the mean and median
mean = jitter.mean()
median = jitter.median()

# Calculate other statistical measures
range_val = jitter.max() - jitter.min()
std_dev = jitter.std()
skewness = jitter.skew()

# # Find the minimum and maximum values
min_jitter = jitter.min()
max_jitter = jitter.max()

# Print the results
print("Min JITTER:", min_jitter)
print("Max JITTER:", max_jitter)
print("Mean JITTER:", mean)
print("Median JITTER:", median)
print("Range JITTER:", range_val)
print("Standard Deviation JITTER:", std_dev)
print("Skewness JITTER:", skewness)

Min JITTER: 0.01
Max JITTER: 4368.64
Mean JITTER: 6.445268932298635
Median JITTER: 1.84
Range JITTER: 4368.63
Standard Deviation JITTER: 45.467535542184905
Skewness JITTER: 66.23153331326267


Descriptive stats for LATENCY for Winnipeg

In [21]:
# Filter the DataFrame to include only rows where CLIENT CITY is Winnipeg
filtered_df = df[df['CLIENT CITY'] == 'Winnipeg']

# Extract the LATENCY column as a Series
latency = df['LATENCY']

# Calculate the mean and median
mean = latency.mean()
median = latency.median()

# Calculate other statistical measures
range_val = latency.max() - latency.min()
std_dev = latency.std()
skewness = latency.skew()

# # Find the minimum and maximum values
min_latency = latency.min()
max_latency = latency.max()

# Print the results
print("Min LATENCY:", min_latency)
print("Max LATENCY:", max_latency)
print("Mean LATENCY:", mean)
print("Median LATENCY:", median)
print("Range LATENCY:", range_val)
print("Standard Deviation LATENCY:", std_dev)
print("Skewness LATENCY:", skewness)

Min LATENCY: 0.07
Max LATENCY: 1542.32
Mean LATENCY: 49.454734412630444
Median LATENCY: 41.99
Range LATENCY: 1542.25
Standard Deviation LATENCY: 37.24061383790003
Skewness LATENCY: 12.83428834482966


Descriptive stats for PACKET LOSS for Winnipeg

In [22]:
# Filter the DataFrame to include only rows where CLIENT CITY is Winnipeg
filtered_df = df[df['CLIENT CITY'] == 'Winnipeg']

# Extract the PACKET LOSS column as a Series
packet_loss = df['PACKET LOSS']

# Calculate the mean and median
mean = packet_loss.mean()
median = packet_loss.median()

# Calculate other statistical measures
range_val = packet_loss.max() - packet_loss.min()
std_dev = packet_loss.std()
skewness = packet_loss.skew()

# # Find the minimum and maximum values
min_packet_loss = packet_loss.min()
max_packet_loss = packet_loss.max()

# Print the results
print("Min PACKET LOSS:", min_packet_loss)
print("Max PACKET LOSS:", max_packet_loss)
print("Mean PACKET LOSS:", mean)
print("Median PACKET LOSS:", median)
print("Range PACKET LOSS:", range_val)
print("Standard Deviation PACKET LOSS:", std_dev)
print("Skewness PACKET LOSS:", skewness)

Min PACKET LOSS: 0.0
Max PACKET LOSS: 47.85
Mean PACKET LOSS: 2.2082398983141553
Median PACKET LOSS: 0.08
Range PACKET LOSS: 47.85
Standard Deviation PACKET LOSS: 4.4839955131340945
Skewness PACKET LOSS: 3.224587642165551


Validating postal codes in Winnipeg against longitude and latitude data

In [23]:
import re

# Function to validate postal codes and check coordinates
def validate_postal_codes():
    invalid_postal_codes = []
    for index, row in df.iterrows():
        postal_code = row['CLIENT POSTAL CODE']
        city = row['CLIENT CITY']

        if city.lower() == 'winnipeg':
            try:
                # Validate the postal code using a regular expression pattern
                pattern = r'^[A-Za-z]\d[A-Za-z]\s?\d[A-Za-z]\d$'
                is_valid = re.match(pattern, postal_code)

                if not is_valid:
                    invalid_postal_codes.append(postal_code)
                else:
                    # Get the expected coordinates for the postal code
                    expected_latitude, expected_longitude = row['GEO LATITUDE'], row['GEO LONGITUDE']

                    # Check if the coordinates match the expected values
                    if expected_latitude != row['GEO LATITUDE'] or expected_longitude != row['GEO LONGITUDE']:
                        invalid_postal_codes.append(postal_code)
            except Exception as e:
                # Error occurred during validation
                print(f"Error: {str(e)}")

    return invalid_postal_codes

# Check and list the invalid postal codes for Winnipeg
invalid_codes = validate_postal_codes()
if invalid_codes:
    print("Invalid postal codes for Winnipeg:")
    for postal_code in invalid_codes:
        print(postal_code)
else:
    print("All postal codes for Winnipeg are valid.")

Invalid postal codes for Winnipeg:
None
None
None
None
R2X2H
R2N
R2N
R2N
R2N
R2N
R2N
R2N


Dropping the invalid Postal Code rows